core/src/main/java/hudson/FilePath.java

```diff
@@ -2008,6 +2021,23 @@ public void copyTo(FilePath target) throws IOException, InterruptedException {
      * @since 1.311
      */
     public void copyToWithPermission(FilePath target) throws IOException, InterruptedException {
+        // Use NIO copy with StandardCopyOption.COPY_ATTRIBUTES when copying on the same machine.
+        if (this.channel == target.channel) {
+            act(new SecureFileCallable<Void>() {
+                public Void invoke(File f, VirtualChannel channel) throws IOException {
+                    try {
+                        File targetFile = new File(target.remote).getAbsoluteFile();
+                        mkdirs(targetFile.getParentFile());
+                        Files.copy(reading(f).toPath(), writing(targetFile).toPath(), StandardCopyOption.COPY_ATTRIBUTES, StandardCopyOption.REPLACE_EXISTING);
```

Comment:
COPY_ATTRIBUTES does preserve setgid/setuid/sticky bits, unlike `setPosixFilePermissions`.
I don't see a likely use case for maintaining the special permission bits when copying across different machines, but on the same machine I think it sense, and this should help prevent regressions for valid use cases of the special bits.